src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+        pkPath = path;
+    }
+
+    void loadProvingKey() {
+        if (!pk) {
+            if (!pkPath) {
+                throw std::runtime_error("proving key path unknown");
+            }
+            loadFromFile(*pkPath, pk);
+        }
+    }
+
+    void saveProvingKey(std::string path) {
+        if (pk) {
+            saveToFile(path, *pk);
+        } else {
+            throw std::runtime_error("cannot save proving key; key doesn't exist");
+        }
+    }
+    void loadVerifyingKey(std::string path) {
+        loadFromFile(path, vk);
+    }
+    void saveVerifyingKey(std::string path) {
+        if (vk) {
+            saveToFile(path, *vk);
+        } else {
+            throw std::runtime_error("cannot save verifying key; key doesn't exist");
+        }
+    }
+
+    void generate() {
+        protoboard<FieldT> pb;
+
+        joinsplit_gadget<FieldT, NumInputs, NumOutputs> g(pb);
+        g.generate_r1cs_constraints();
+
+        const r1cs_constraint_system<FieldT> constraint_system = pb.get_constraint_system();
+        r1cs_ppzksnark_keypair<ppzksnark_ppT> keypair = r1cs_ppzksnark_generator<ppzksnark_ppT>(constraint_system);
+
+        pk = keypair.pk;
+        vk = keypair.vk;
+    }
+
+    JoinSplitCircuit() {}
+
+    bool verify(
+        const std::string& proof,
+        const uint256& pubKeyHash,
+        const uint256& randomSeed,
+        const boost::array<uint256, NumInputs>& macs,
+        const boost::array<uint256, NumInputs>& nullifiers,
+        const boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!vk) {
+            throw std::runtime_error("JoinSplit verifying key not loaded");
+        }
+
+        r1cs_ppzksnark_proof<ppzksnark_ppT> r1cs_proof;
+        std::stringstream ss;
+        ss.str(proof);
+        ss >> r1cs_proof;
+
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
+
+        auto witness = joinsplit_gadget<FieldT, NumInputs, NumOutputs>::witness_map(
+            rt,
+            h_sig,
+            macs,
+            nullifiers,
+            commitments,
+            vpub_old,
+            vpub_new
+        );
+
+        return r1cs_ppzksnark_verifier_strong_IC<ppzksnark_ppT>(*vk, witness, r1cs_proof);
+    }
+
+    std::string prove(
+        const boost::array<JSInput, NumInputs>& inputs,
+        const boost::array<JSOutput, NumOutputs>& outputs,
+        boost::array<Note, NumOutputs>& output_notes,
+        boost::array<ZCNoteEncryption::Ciphertext, NumOutputs>& ciphertexts,
+        uint256& ephemeralKey,
+        const uint256& pubKeyHash,
+        uint256& randomSeed,
+        boost::array<uint256, NumInputs>& macs,
+        boost::array<uint256, NumInputs>& nullifiers,
+        boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
```

Comment:
If it's `const` it's an input, if it's not it's an output, with the exception of `vpub_old` and `vpub_new`.